workers/unity/Packages/com.improbable.gdk.core/Systems/CleanReactiveComponentsSystem.cs

```diff
@@ -54,32 +53,40 @@ private void GenerateComponentGroups()
                         continue;
                     }
 
-                    if (typesToRemove.Contains(type))
+                    if (typesToRemove.Add(type))
                     {
-                        continue;
+                        addRemoveComponentActionMethod.MakeGenericMethod(type).Invoke(this, null);
                     }
-
-                    typesToRemove.Add(type);
-                    addRemoveComponentActionMethod.MakeGenericMethod(type).Invoke(this, null);
                 }
 
                 // Find all ComponentCleanupHandlers
                 foreach (var type in assembly.GetTypes())
                 {
                     if (typeof(ComponentCleanupHandler).IsAssignableFrom(type) && !type.IsAbstract)
                     {
+                        // These are generic and cannot use `RemoveAtEndOfTick`
                         var componentCleanupHandler = (ComponentCleanupHandler) Activator.CreateInstance(type);
                         foreach (var componentType in componentCleanupHandler.CleanUpComponentTypes)
                         {
                             typesToRemove.Add(componentType.GetManagedType());
                             addRemoveComponentActionMethod.MakeGenericMethod(componentType.GetManagedType()).Invoke(this, null);
                         }
 
+                        // EventsReceived groups
+                        var eventGroups = new ComponentGroup[componentCleanupHandler.EventComponentTypes.Length];
+                        for (var i = 0; i < componentCleanupHandler.EventComponentTypes.Length; i++)
+                        {
+                            var componentType = componentCleanupHandler.EventComponentTypes[i];
+                            eventGroups[i] = GetComponentGroup(componentType);
+                        }
+
+                        // Updates group
                         componentCleanups.Add(new ComponentCleanup
                         {
                             Handler = componentCleanupHandler,
                             UpdateGroup = GetComponentGroup(componentCleanupHandler.ComponentUpdateType),
-                            AuthorityChangesGroup = GetComponentGroup(componentCleanupHandler.AuthorityChangesType)
+                            AuthorityChangesGroup = GetComponentGroup(componentCleanupHandler.AuthorityChangesType),
+                            EventGroups = eventGroups,
```

Comment:
Can LINQ this. 

`EventGroups = componentCleanupHandler.EventComponentTypes.Select(t => GetComponentGroup(t))`

Reply:
Will do